src/test/java/htsjdk/samtools/cram/CarrotTest.java

```diff
@@ -0,0 +1,191 @@
+package htsjdk.samtools.cram;
+
+import htsjdk.HtsjdkTest;
+import htsjdk.beta.plugin.IOUtils;
+import htsjdk.io.HtsPath;
+import htsjdk.io.IOPath;
+import htsjdk.samtools.SAMRecord;
+import htsjdk.samtools.SamReader;
+import htsjdk.samtools.SamReaderFactory;
+import htsjdk.samtools.ValidationStringency;
+import htsjdk.samtools.cram.build.CramIO;
+import htsjdk.samtools.cram.common.CramVersions;
+import htsjdk.samtools.util.FileExtensions;
+import htsjdk.utils.SamtoolsTestUtils;
+import org.testng.Assert;
+import org.testng.annotations.DataProvider;
+import org.testng.annotations.Test;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.file.Path;
+import java.util.Iterator;
+
+// A Command line program for large-scale Carrot round-trip CRAM tests, plus some tests for the command line app.
+public class CarrotTest extends HtsjdkTest {
+    private static String CONVERT_USAGE = "convert reads_file reference samtools_view_params";
+    private static String COMPARE_USAGE = "compare reads_file1 reads_file2 reference";
+    final static int EXIT_CODE_SUCCESS = 0;
+    final static int EXIT_CODE_DIFFS = 1;
+    final static int EXIT_CODE_FAILURE = -1;
+
+    // supported CLP commands - the string representation of these values are the actual command strings
+    public enum CLPCommand {
+        CONVERT,        // convert an input to CRAM 3.1 using samtools
+        COMPARE         // compare two crams for fidelity
+    };
+
+    // 1) convert to CRAM 3.1 using samtools:
+    //  inputs:
+    //      convert
+    //      input reads file (bam or cram)
+    //      output reads file
+    //      reference
+    //      samtools params
+    // 2) compare:
+    //  inputs:
+    //      compare
+    //      reads file 1 (bam or cram)
+    //      reads file 2 (bam or cram)
+    //      reference
+    public static void main(final String[] args) {
```

Comment:
This feels awkwardly placed but doing it differently seems like more of a hassle.  I think you've come up with a pretty good solution.

Reply:
I'm actually reverting the carrot changes in this branch (reverting the one commit) since its not complete/working anyway. I'll revive it for the write branch though.